PICMI_Python/particles.py

```diff
@@ -203,402 +245,401 @@ def __getitem__(self, key):
+                if i < len(momentum_exprs) and momentum_exprs[i] is not None:
+                    momentum_exprs[i] = f'{momentum_exprs[i]}'.replace('\n', '')
+        
+        # Normalize momentum spread expressions
+        if 'momentum_spread_expressions' in data and data['momentum_spread_expressions']:
+            momentum_spread_exprs = data['momentum_spread_expressions']
+            for i in range(3):
+                if i < len(momentum_spread_exprs) and momentum_spread_exprs[i] is not None:
+                    momentum_spread_exprs[i] = f'{momentum_spread_exprs[i]}'.replace('\n', '')
+        
+        return data
+    
+    @model_validator(mode='after')
+    def _convert_expressions_and_extract_kw(self) -> Self:
+        """Convert momentum expressions to strings and extract user-defined keywords"""
+        # Convert momentum expressions to string if needed
         for idir in range(3):
             if self.momentum_expressions[idir] is not None:
                 self.momentum_expressions[idir] = f'{self.momentum_expressions[idir]}'.replace('\n', '')
             if self.momentum_spread_expressions[idir] is not None:
                 self.momentum_spread_expressions[idir] = f'{self.momentum_spread_expressions[idir]}'.replace('\n', '')
-
-        # --- Find any user defined keywords in the kw dictionary.
-        # --- Save them and delete them from kw.
-        # --- It's up to the code to make sure that all parameters
-        # --- used in the expression are defined.
-        self.user_defined_kw = {}
-        for k in list(kw.keys()):
-            if re.search(r'\b%s\b'%k, self.density_expression):
-                self.user_defined_kw[k] = kw[k]
-                del kw[k]
-            elif self.momentum_expressions[0] is not None and re.search(r'\b%s\b'%k, self.momentum_expressions[0]):
-                self.user_defined_kw[k] = kw[k]
-                del kw[k]
-            elif self.momentum_expressions[1] is not None and re.search(r'\b%s\b'%k, self.momentum_expressions[1]):
-                self.user_defined_kw[k] = kw[k]
-                del kw[k]
-            elif self.momentum_expressions[2] is not None and re.search(r'\b%s\b'%k, self.momentum_expressions[2]):
-                self.user_defined_kw[k] = kw[k]
-                del kw[k]
-
-        self.handle_init(kw)
+        
+        # Extract user-defined keywords from extra fields
+        if self.model_extra:
+            self.user_defined_kw = {}
+            keys_to_remove = []
+            
+            # Check density expression
+            for k, v in self.model_extra.items():
+                if re.search(rf'\b{k}\b', self.density_expression):
+                    self.user_defined_kw[k] = v
+                    keys_to_remove.append(k)
+                # Check momentum expressions
+                elif self.momentum_expressions[0] is not None and re.search(rf'\b{k}\b', self.momentum_expressions[0]):
+                    self.user_defined_kw[k] = v
```

Comment:
See above. Don´t think this is still needed, we can derive in downstream codes and add new fields with their own validators/types.